File: app/src/main/java/com/murrayde/animekingmobile/view/community/data_source/MainActivityViewModel.java
package com.murrayde.animekingmobile.view.community.data_source;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

import com.murrayde.animekingmobile.di.community.AnimeApiComponent;
import com.murrayde.animekingmobile.di.community.DaggerAnimeApiComponent;
import com.murrayde.animekingmobile.network.community.AnimeApiEndpoint;
import com.murrayde.animekingmobile.network.community.api.AnimeData;
import com.murrayde.animekingmobile.util.PagingUtil;

import io.reactivex.disposables.CompositeDisposable;

public class MainActivityViewModel extends ViewModel {

    private CompositeDisposable compositeDisposable;

    private LiveData<PagedList<AnimeData>> animeData;
    private MutableLiveData<Boolean> networkDoneLoading;

    public MainActivityViewModel() {
        AnimeApiComponent daggerAnimeApiComponent = DaggerAnimeApiComponent.builder()
                .build();
        AnimeApiEndpoint animeApiEndpoint = daggerAnimeApiComponent.getAnimeApiEndpoint();

        compositeDisposable = new CompositeDisposable();
        networkDoneLoading = new MutableLiveData<Boolean>();
        AnimeDataSourceFactory animeDataSourceFactory = new AnimeDataSourceFactory(animeApiEndpoint, compositeDisposable, networkDoneLoading);

        PagedList.Config pagingConfig = new PagedList.Config.Builder()
                .setPrefetchDistance(PagingUtil.PAGING_PREFETCH)
                .setEnablePlaceholders(false)
                .build();
        animeData = new LivePagedListBuilder<>(animeDataSourceFactory, pagingConfig).build();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        compositeDisposable.clear();
    }

    public LiveData<PagedList<AnimeData>> getAnimeData() {
        return animeData;
    }

    public LiveData<Boolean> networkDoneLoading() {
        return networkDoneLoading;
    }
}

File: app/src/main/java/com/murrayde/animekingmobile/view/community/data_source/MainActivityViewModel.kt
package com.murrayde.animekingmobile.view.community.data_source

import android.util.Log
import androidx.hilt.lifecycle.ViewModelInject
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.murrayde.animekingmobile.model.ui.AnimeForYou
import com.murrayde.animekingmobile.repository.community.MainRepo
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableSingleObserver
import io.reactivex.schedulers.Schedulers
import timber.log.Timber

class MainActivityViewModel @ViewModelInject constructor(private val mainRepo: MainRepo) : ViewModel() {

    private val disposable = CompositeDisposable()
    private val animeForYou = MutableLiveData<AnimeForYou>()
    private val loading = MutableLiveData<Boolean>()

    init {
        fetchAnimeForYou()
    }

    private fun fetchAnimeForYou() {
        loading.postValue(true)
        disposable.add(
                mainRepo.getAnimeForYou().subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(object : DisposableSingleObserver<AnimeForYou>() {
                            override fun onSuccess(t: AnimeForYou) {
                                loading.postValue(false)
                                animeForYou.value = t
                                Log.d(MainActivityViewModel::class.qualifiedName, "Data retrieved")
                            }

                            override fun onError(e: Throwable) {
                                loading.postValue(false)
                                Timber.e("Error loading titles")
                                Log.e(MainActivityViewModel::class.qualifiedName, "Data not retrieved")

                            }

                        })
        )
    }

    fun getAnimeForYou(): LiveData<AnimeForYou> = animeForYou

    val networkLoading: LiveData<Boolean> = loading

    override fun onCleared() {
        super.onCleared()
        disposable.clear()
    }
}